pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -162,15 +178,41 @@ func (c *Controller) processPVC(ctx context.Context, pvcNamespace, pvcName strin
 	}
 
 	if protectionutil.IsDeletionCandidate(pvc, volumeutil.PVCProtectionFinalizer) {
-		// PVC should be deleted. Check if it's used and remove finalizer if
-		// it's not.
-		isUsed, err := c.isBeingUsed(ctx, pvc)
+		// PVC should be deleted. Check if it's used and remove finalizer if it's not.
+		isUsed, err := c.isBeingUsed(pvc)
 		if err != nil {
 			return err
 		}
 		if !isUsed {
-			return c.removeFinalizer(ctx, pvc)
+			// PVC is not used by any Pods from Informer. Considering the objects in Informer
+			// might not be consistent with APIServer yet, we introduce retries with backoff
+			// to check the usage multiple times.
+			var retryCount int
+			if val, ok := c.pvcUsedCheckRetryingCache.Load(pvcKey); ok {
+				retryCount = val.(int)
+			}
+			klog.V(4).InfoS("Check terminating PVC not being used from Informer",
+				"retries", fmt.Sprintf("%d/%d", retryCount, maxUsedCheckRetries),
+				"PVC", klog.KRef(pvcNamespace, pvcName))
+
+			// If it has not reached the maxUsedCheckRetries, put the PVC into delaying queue with backoff.
+			if retryCount < maxUsedCheckRetries {
+				c.pvcUsedCheckRetryingCache.Store(pvcKey, retryCount+1)
+				c.queue.AddAfter(pvcKey, defaultUsedCheckBackOff)
+				return nil
+			}
+
+			if err = c.removeFinalizer(ctx, pvc); err == nil {
+				// The PVC in cache should only be deleted when removeFinalizer succeeds,
+				// otherwise it will lead to duplicate retries when removeFinalizer fails.
+				c.pvcUsedCheckRetryingCache.Delete(pvcKey)
```

Comment:
Maybe we should just move it to `removeFinalizer`?

Reply:
Yeah, we can do that. But it seems to be more complicated if we use `pvcUsedCheckRetryingCache` in multiple different places, and we also have to add `pvcKey` to `removeFinalizer` args. Do you think it is more understandable to only read/write/delete `pvcUsedCheckRetryingCache` in this PVC deletion code block?

Reply:
Hm, initially I was thinking that it can be easy to miss that we should delete the key from `c.pvcUsedCheckRetryingCache`, if the code will change in the future - we don't want to let the cache grow into infinity. And having that in the `c.removeFinalizer` would remove the need to remember about that. But right now I'm having second thoughts and it doesn't sound as good as in the beginning :P So we can just leave it.